openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+    /// tree is not mutable and thus cannot be pruned, a lightweight mask tree
+    /// with the same topology is created that can be pruned to use as a
+    /// reference. Initialization of this mask tree can optionally be disabled
+    /// for delayed construction.
+    TreeToMerge(const TreeType* constTree, DeepCopy, bool initialize = true)
+        : mConstTree(constTree)
+    {
+        if (mConstTree && initialize)     this->initializeMask();
+    }
+
+    /// @brief Non-const tree pointer constructor for deep-copying data. The
+    /// tree is not intended to be modified so is not pruned, instead a
+    /// lightweight mask tree with the same topology is created that can be
+    /// pruned to use as a reference. Initialization of this mask tree can
+    /// optionally be disabled for delayed construction.
+    TreeToMerge(TreeType* tree, DeepCopy tag, bool initialize = true)
+        : TreeToMerge(static_cast<const TreeType*>(tree), tag, initialize) { }
+
+    /// @brief Reset the non-const tree shared pointer. This is primarily
+    /// used to preserve the order of trees to merge in a container but have
+    /// the data in the tree be lazily loaded or resampled.
+    void reset(typename TreeType::Ptr treePtr)
+    {
+        mTreePtr = treePtr;
+        mTree = mTreePtr.get();
+        mConstTree = nullptr;
+    }
+
+    /// @brief Return a pointer to the non-const tree.
+    TreeType* tree() { return mTree; }
+    /// @brief Return a pointer to the const tree.
+    const TreeType* constTree() { return mConstTree; }
+
+    /// @brief Retrieve a const pointer to the root node.
+    const RootNodeType* rootPtr() const;
+
+    /// @brief Return a pointer to the node of type @c NodeT that contains
+    /// voxel (x, y, z).  If no such node exists, return @c nullptr.
+    template<typename NodeT>
+    const NodeT* probeConstNode(const Coord& ijk) const;
+
+    /// @brief Return a pointer to the node of type @c NodeT that contains voxel (x, y, z).
+    /// If the tree is non-const, steal the node and replace it with an inactive
+    /// background-value tile.
+    /// If the tree is const, deep-copy the node and modify the mask tree to prune the node.
+    template <typename NodeT>
+    std::unique_ptr<NodeT> stealOrDeepCopyNode(const Coord& ijk);
+
+    /// @brief Add a tile containing voxel (x, y, z) at the level of NodeT,
+    /// deleting the existing branch if necessary.
+    template <typename NodeT>
+    void addTile(const Coord& ijk, const ValueType& value, bool active);
+
+    // build a lightweight mask using a union of the const tree where leaf nodes
+    // are converted into active tiles
+    void initializeMask();
+
+    // returns true if mask has been initialized
+    bool hasMask() const;
+
+private:
+    struct MaskUnionOp;
+
+    typename TreeType::Ptr mTreePtr;
+    TreeType* mTree = nullptr;
+    const TreeType* mConstTree = nullptr;
+    typename MaskTreeType::Ptr mMaskTree;
```

Comment:
Personally I'm not a fan of a single class representing two states of constness. I think a better design would be for a specialisation of this convenience class which is templated on a const type. `stealOrDeepCopyNode` could then be called something like `getNode` and be defined with respect to the specialisation - this also better matches the naming of `addTile` with respect to its behaviour. It also removes this tagging constructor stuff (though you could still keep it to ensure the correct specialisation has been selected) and simplifies the members (no mask on non-const). That would be my preference as this class is pretty much boilerplate except for those methods. This wouldn't be as much of an issue if this wasn't part of the public API and only used internally. Also, do you need to store a shared pointer to the tree here? Could these be references instead and remove the ability to pass nullptrs? Could the mask be a `unique_ptr` or have I misunderstood how that's used?

Reply:
So there'd be two TreeToMerge specializations:

TreeToMergeDeepCopy
TreeToMergeSteal

The Deep copy would take a const pointer, and the Steal would take a non-const?

Won't this run into the same issue where we'd still need something to represent *either* type of tree?  For when we have a std::vector<TreeToMerge> and want it to have a mix of deepcopy and steal trees?  We're back to a virtual base class; or building another opaque wrapper around TreeToMergeDeepCopy and TreeToMergeSteal to alias them...  Since this is rather special case and lightweight, I don't see anything wrong with just collapsing it into a single structure.

Here's a similar class we ended up with in the HDK, FWIW:
https://www.sidefx.com/docs/hdk16.0/class_g_o_p___group_parse_1_1_group_creator.html

Reply:
Thinking on it, it isn't the case that TreeToMerge is "Const" or "Non-Const".  It is either "Stealable" or "Deep-Copy"

I'd be rather tempted to just const_cast away the const inside of TreeToMerge so it only needs a single tree pointer type, but instead uses a state flag if it is in DeepCopy or Steal mode.

Reply:
> The Deep copy would take a const pointer, and the Steal would take a non-const?
> Won't this run into the same issue where we'd still need something to represent either type of tree?  For when we have a std::vector and want it to have a mix of deepcopy and steal trees?

I think we'd just have two vectors, one for each specialization - no need for a shared base or any more infrastructure

Reply:
> I think we'd just have two vectors, one for each specialization - no need for a shared base or any more infrastructure

I think this a point I should make clearer in the documentation. It is _really_ important to preserve the order of grids regardless of whether or not they are deep-copy or steal. The reason is that at least doing certain operations over floating-point trees, you'll get different results due to floating-point associativity. I believe strongly that there should never be a change in the resulting grid by changing how you're pulling data from it (ie stealing instead of deep-copying), so we need to mix the two types while preserving the order. I tried a few different implementations of this such as using two vectors as you're suggesting @Idclip and simply storing an integer parameter for the order. I found that it ended up with a more complicated public API and was more complicated to manage internally too, so I settled on this wrapped tree approach where the deep-copy/steal state was an abstraction.

> I'd be rather tempted to just const_cast away the const inside of TreeToMerge so it only needs a single tree pointer type, but instead uses a state flag if it is in DeepCopy or Steal mode.

I think that's a good suggestion. As long as there's no way to pass in a const tree to be stolen, there should be no opportunity for abuse. I'll try that out.

> Also, do you need to store a shared pointer to the tree here?

I found this useful when I was resampling from an existing const tree so as to match the target transform (a very common operation!). In that case, you want to be able to steal from this new temporary tree, but you also need to hold on to a shared pointer somewhere to maintain ownership. It didn't seem like a major extension to this class to support this and cleaned up the client code quite a bit not having to worry about storing a buffer of these temporary trees.

> Could these be references instead and remove the ability to pass nullptrs?

It's quite convenient to be able to store a TreeToMerge class with a nullptr because you can use it as a placeholder for where a tree will be inserted. This means you can do things asynchoronously - loading a tree or resampling another tree as discussed above.

> Could the mask be a unique_ptr or have I misunderstood how that's used?

Yes, the mask can be a unique pointer, I can change that. It would be nice to have a unique pointer typedef or using alias in the Tree class.

Reply:
>  I tried a few different implementations of this such as using two vectors as you're suggesting @Idclip and simply storing an integer parameter for the order. I found that it ended up with a more complicated public API and was more complicated to manage internally too, so I settled on this wrapped tree approach where the deep-copy/steal state was an abstraction.

Fair enough - I would say however that I'm not sure how this would create a more complicated **public** API, unless you do mean the need for this index tracking. You reference fp arithmetic which is absolutely a valid case for retaining order, but ops like CsgUnion are a great example where this does not need to be respected. You're providing a container which can be utilised within a custom visitor, but it's down to you to implement this visitor correctly - and, if by correctly we mean must adhere to deterministic ordering, then that's something you as a user must do. Having a single container just so it can be stored in a single vector to help with this feels a bit arbitrary, especially along with the other issues identified. I'll abstain from any further comments on this though.

> In that case, you want to be able to steal from this new temporary tree, but you also need to hold on to a shared pointer somewhere to maintain ownership.
>  It didn't seem like a major extension to this class to support this and cleaned up the client code quite a bit not having to worry about storing a buffer of these temporary trees.

Right so you're using this container for new storage too? When you say buffer of tmp trees, this could be achieved with another vector of tree ptrs? Removing the ability for this class to hold a null tree imo would be a worthwhile thing to consider (you'd remove a bunch of if(tree) checks too). Along with "Pay for what you use" etcetc

Reply:
> Fair enough - I would say however that I'm not sure how this would create a more complicated public API, unless you do mean the need for this index tracking. 

Essentially, yes you need to track the index on the client end to these types of operations. Where this complicates a public API is for asynchronous workflows such as the resampling example and where you care about order. In this case, you want to swap the contents of one of these TreeToMerge objects with a new tree. To do so, you have to maintain a reference to this TreeToMerge object or if you use two tree vectors, an index and a flag to select the right element from the right vector. 

> You reference fp arithmetic which is absolutely a valid case for retaining order, but ops like CsgUnion are a great example where this does not need to be respected. You're providing a container which can be utilised within a custom visitor, but it's down to you to implement this visitor correctly - and, if by correctly we mean must adhere to deterministic ordering, then that's something you as a user must do.

It is a valid point and I did consider this. On balance, I think maintaining a consistency of APIs to all multi-tree operators is nice. That means not having a CsgUnionOp with one set of constructors and a FogVolumeAddOp (for example) with a different set of constructors because it needs to be deterministic. Internally, we could extract the data from one of these TreeToMerge arrays on construction if we know determinism isn't required, but the benefits in doing so are (relatively) limited and this change won't affect the public API.

> Right so you're using this container for new storage too? When you say buffer of tmp trees, this could be achieved with another vector of tree ptrs? 

Yes, this could be achieved with another vector of tree ptrs. That was where I started with this. However, as you can construct a TreeToMerge array _before_ constructing the operator, you'd end up having to store this vector of tree pointers in the client code. It is a convenience not to have to do this. If you'll still thinking about this along the lines of doing away with the TreeToMerge array for the CsgUnionOp, then consider it in the context of a deterministic operator such as FogVolumeAddOp.

> Removing the ability for this class to hold a null tree imo would be a worthwhile thing to consider (you'd remove a bunch of if(tree) checks too). Along with "Pay for what you use" etcetc

This is probably a tiny optimization. Nonetheless, I will try and get rid of the ability to store a nullptr - with the slight change to the internal storage, that might be more feasible. A related question is not being able to optimize away the check if a tree can be stolen from or not in the inner loop, despite being able to trivially check if _any_ tree in the array can be stolen from. I think the simple solution to this is to tweak the implementation of the three methods on the critical path to add a new template parameter:

```
    template<typename NodeT, bool DeepCopyOnly = false>
    const NodeT* probeConstNode(const Coord& ijk) const;
```

That way, you could modify the internals of the operator to take advantage of this. I'm not going to do that here, just highlighting that these kind of optimizations could theoretically be introduced without breaking the public API. Let's not lose sight of the huge performance boost in using this technique in the first place. :)

Reply:
> This is probably a tiny optimization.
> Let's not lose sight of the huge performance boost in using this technique in the first place. :)

Absolutely - this feedback isn't related to performance optimizations (i.e. removing the nullptr check is imo a better overall design). Feel free to use or discard these points as you see fit!

Reply:
> Yes, the mask can be a unique pointer, I can change that. It would be nice to have a unique pointer typedef or using alias in the Tree class.

We ran into this ourselves, as we often have a Ptr typedef.  In our case Ptr is supposed to be a raw pointer, so we use SPtr and UPtr as the defaults for shared and unique pointers.  Personally I don't like Ptr being an implicit shared pointer, as there are performance and ownership implications that should be be made explicit.

Reply:
Yes, it's not ideal having Ptr being a shared pointer. I'm not sure how we'd go about changing that though, it would break so much code! One tricky aspect of using a unique pointer as a member of a class is not being able to implicitly define the copy constructor. In this case, I wrapped in it a little struct that only contains the unique pointer and does a deep-copy on copy construction so that it's not necessary to add logic for copying all the other member variables.